# eBay Car Sales Data

We are going to explore the clean the ebay car sales data and analyze 

In [1]:
import pandas as pd
autos=pd.read_csv('autos.csv',encoding='Latin-1')

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
autos.columns=['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'year_of_registration', 'gearbox', 'power_ps', 'model',
       'odometer', 'month_of_registration', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']

In [7]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  year_of_registration  \
0  privat    Angebot  $5,000  control          bus                  2004   
1  privat    Angebot  $8,500  control    limousine                  1997   
2  privat    Angebot  $8,990     test    limousine                  2009   
3  privat    Angebot  $4,350  control   kleinwagen                  2007   
4  privat    Angebot  $1,350     test        kombi                  2003   

     gearbox  power_ps   model   odometer  month_of_registration fuel_type  \
0    manuell       158  andere  150,000km                      3       lpg   
1  automatik       286     7er  150,000km                      6    benzin   
2    manuell       102    golf   70,000km                      7    benzin   
3  automatik        71  fortwo   70,000km                      6    benzin   
4    manuell         0   focus  150,000km                      7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

We made the above changes to the column names to make them more descriptive as well as easier to use for further operations

In [8]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  year_of_registration  gearbox      power_ps  model  \
count         44905          50000.000000    47320  50000.000000  47242   
unique            8                   NaN        2           NaN    245   
top       limousine                   NaN  manuell           NaN   golf   
freq          12859                   NaN    36993           NaN   4024   
mean            NaN           2005.073280      NaN    116.355920    NaN   
std             NaN            105.712813      NaN    209.216627    NaN   
min             NaN           1000.000000      NaN      0.000000    NaN   
25%             NaN           1999.000000      NaN     70.000000    NaN   
50%             NaN           2003.000000      NaN    105.000000    NaN   
75%             NaN           2008.000000      NaN    150.000000    NaN   
max             NaN           9999.000000      NaN  17700.000000    NaN   

         odometer  month_of_registration fuel_type       brand  \
count       50000           50000.000000     45518       50000   
unique         13                    NaN         7          40   
top     150,000km                    NaN    benzin  volkswagen   
freq        32424                    NaN     30107       10687   
mean          NaN               5.723360       NaN         NaN   
std           NaN               3.711984       NaN         NaN   
min           NaN               0.000000       NaN         NaN   
25%           NaN               3.000000       NaN         NaN   
50%           NaN               6.000000       NaN         NaN   
75%           NaN               9.000000       NaN         NaN   
max           NaN              12.000000       NaN         NaN   

       unrepaired_damage           ad_created  num_photos   postal_code  \
count              40171                50000     50000.0  50000.000000   
unique                 2                   76         NaN           NaN   
top                 nein  2016-04-03 00:00:00         NaN           NaN   
freq               35232                 1946         NaN           NaN   
mean                 NaN                  NaN         0.0  50813.627300   
std                  NaN                  NaN         0.0  25779.747957   
min                  NaN                  NaN         0.0   1067.000000   
25%                  NaN                  NaN         0.0  30451.000000   
50%                  NaN                  NaN         0.0  49577.000000   
75%                  NaN                  NaN         0.0  71540.000000   
max                  NaN                  NaN         0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

1. Initial observations are that the columns 'seller' and 'offer_type' have almost all the entries equal
2. The 'num_photos' column also looks odd, so we will investigate it further


In [9]:
autos['num_photos'].value_counts()

0    50000
Name: num_photos, dtype: int64

It seems that all the values in the 'num_photos' column are zero hence we are going to drop it along with the 'seller' and 'offer_type'

In [10]:
autos.drop(['seller' ,'offer_type', 'num_photos'],axis=1)
autos.head(2)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   

  offer_type   price   abtest vehicle_type  year_of_registration    gearbox  \
0    Angebot  $5,000  control          bus                  2004    manuell   
1    Angebot  $8,500  control    limousine                  1997  automatik   

   power_ps   model   odometer  month_of_registration fuel_type    brand  \
0       158  andere  150,000km                      3       lpg  peugeot   
1       286     7er  150,000km                      6    benzin      bmw   

  unrepaired_damage           ad_created  num_photos  postal_code  \
0              nein  2016-03-26 00:00:00           0        79588   
1              nein  2016-04-04 00:00:00           0        71034   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08

Upon further inspection into the df we notice that the values in the price and odometer section are stored as strings, so we are going to clean them are store them as float and are going to change the column name appropriately

In [11]:
autos['price']=(autos['price'].str.replace('$','')
                                .str.replace(',','')
                                .astype(float))
autos['odometer']=(autos['odometer'].str.replace('km','')
                                .str.replace(',','')
                                .astype(float))

C:\Users\kvive\AppData\Local\Temp\ipykernel_9620\64574181.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  autos['price']=(autos['price'].str.replace('$','')


In [12]:

autos.rename({'odometer':'odometer_km'},axis=1,inplace=True)
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type    price   abtest vehicle_type  year_of_registration  \
0      privat    Angebot   5000.0  control          bus                  2004   
1      privat    Angebot   8500.0  control    limousine                  1997   
2      privat    Angebot   8990.0     test    limousine                  2009   
3      privat    Angebot   4350.0  control   kleinwagen                  2007   
4      privat    Angebot   1350.0     test        kombi                  2003   
...       ...        ...      ...      ...          ...                   ...   
49995  privat    Angebot  24900.0  control    limousine                  2011   
49996  privat    Angebot   1980.0  control       cabrio                  1996   
49997  privat    Angebot  13200.0     test       cabrio                  2014   
49998  privat    Angebot  22900.0  control        kombi                  2013   
49999  privat    Angebot   1250.0  control    limousine                  1996   

         gearbox  power_ps   model  odometer_km  month_of_registration  \
0        manuell       158  andere     150000.0                      3   
1      automatik       286     7er     150000.0                      6   
2        manuell       102    golf      70000.0                      7   
3      automatik        71  fortwo      70000.0                      6   
4        manuell         0   focus     150000.0                      7   
...          ...       ...     ...          ...                    ...   
49995  automatik       239      q5     100000.0                      1   
49996    manuell        75   astra     150000.0                      5   
49997  automatik        69     500       5000.0                     11   
49998    manuell       150      a3      40000.0                     11   
49999    manuell       101  vectra     150000.0                      1   

      fuel_type       brand unrepaired_damage           ad_created  \
0           lpg     peugeot              nein  2016-03-26 00:00:00   
1        benzin         bmw              nein  2016-04-04 00:00:00   
2        benzin  volkswagen              nein  2016-03-26 00:00:00   
3        benzin       smart              nein  2016-03-12 00:00:00   
4        benzin        ford              nein  2016-04-01 00:00:00   
...         ...         ...               ...                  ...   
49995    diesel        audi              nein  2016-03-27 00:00:00   
49996    benzin        opel              nein  2016-03-28 00:00:00   
49997    benzin        fiat              nein  2016-04-02 00:00:00   
49998    diesel        audi              nein  2016-03-08 00:00:00   
49999    benzin        opel              nein  2016-03-13 00:00:00   

       num_photos  postal_code            last_seen  
0               0        79588  2016-04-06 06:45:54  
1               0        71034  2016-04-06 14:45:08  
2               0        35394  2016-04-06 20:15:37  
3               0        33729  2016-03-15 03:16:28  
4               0        39218  2016-04-01 14:38:50  
...           

Continuing with the exploration we are going to check if we have any outlier or unrealistic in the price and odometer column

In [13]:
autos['odometer_km'].value_counts()

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
5000.0        967
40000.0       819
30000.0       789
20000.0       784
10000.0       264
Name: odometer_km, dtype: int64

We can see that the values are rounded to preset values(maybe by the website)

Also, we can notice that there are more high mileage cars than low mileage cars

In [14]:
print(autos['price'].unique().shape)
print(autos['price'].describe())
print(autos['price'].value_counts().head(20))

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64
0.0       1421
500.0      781
1500.0     734
2500.0     643
1000.0     639
1200.0     639
600.0      531
800.0      498
3500.0     498
2000.0     460
999.0      434
750.0      433
900.0      420
650.0      419
850.0      410
700.0      395
4500.0     394
300.0      384
2200.0     382
950.0      379
Name: price, dtype: int64


We can clearly notice that are some problems with these columns such that some entries are 0 in the price column which is infeasible, given these values only constitute about 2% of the entries we might want to drop these rows totally.

Also, there are 2357 unique values which might suggest that rounding was not the work of website but instead people themselves rounding them off

In [15]:
autos['price'].value_counts().sort_index(ascending=False).head(20)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
Name: price, dtype: int64

In [16]:
autos['price'].value_counts().sort_index(ascending=True).head(10)

0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
Name: price, dtype: int64

We can notice that the values in price column range from **0 till 99999999** which seems like an absurd price for a car.
But we need to consider it is an auction site the prices may be anything. Upon further inspection we find out that the values increase steadily unto $ 350,000 and after that it skyrockets, for analysis’s sake we are going to consider only values in the range $1 to $ 350,000, and we are going to drop other rows


In [17]:
autos=autos[autos['price'].between(1,350000)]
# above we used boolean indexing 

In [18]:
autos['price'].describe(include='all')

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

Next we are going to check the following columns
1. `date_crawled`
2. `registration_month`
3. `registration_year`
4. `ad_created`
5. `last_seen`


In [19]:
autos[['date_crawled','ad_created','last_seen']][:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [20]:
(autos["date_crawled"]
        .str[:10]
        .value_counts( dropna=False)
        .sort_index()
        )

2016-03-05    1230
2016-03-06     682
2016-03-07    1749
2016-03-08    1617
2016-03-09    1607
2016-03-10    1563
2016-03-11    1582
2016-03-12    1793
2016-03-13     761
2016-03-14    1775
2016-03-15    1665
2016-03-16    1438
2016-03-17    1536
2016-03-18     627
2016-03-19    1689
2016-03-20    1840
2016-03-21    1815
2016-03-22    1602
2016-03-23    1565
2016-03-24    1425
2016-03-25    1535
2016-03-26    1564
2016-03-27    1510
2016-03-28    1693
2016-03-29    1656
2016-03-30    1636
2016-03-31    1546
2016-04-01    1636
2016-04-02    1723
2016-04-03    1875
2016-04-04    1772
2016-04-05     636
2016-04-06     154
2016-04-07      68
Name: date_crawled, dtype: int64

this project is on hold to wait for an answer on stackoverflow
the answer has been received and the .str[:10] includes the dates only from the date crawled column.

We may have also used normalize=True attribute in the value_counts function

In [21]:
(autos["date_crawled"]
 .str[:10]
 .value_counts( dropna = False, normalize = True)
 .sort_index()
 )

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

We can clearly see that the crawler crawled the data daily for about one month i.e. March 2016 to April 2016

In [22]:
(autos["date_crawled"]
 .str[:10]
 .value_counts( dropna = False, normalize = True)
 .sort_values()
 )

2016-04-07    0.001400
2016-04-06    0.003171
2016-03-18    0.012911
2016-04-05    0.013096
2016-03-06    0.014043
2016-03-13    0.015670
2016-03-05    0.025327
2016-03-24    0.029342
2016-03-16    0.029610
2016-03-27    0.031092
2016-03-25    0.031607
2016-03-17    0.031628
2016-03-31    0.031834
2016-03-10    0.032184
2016-03-26    0.032204
2016-03-23    0.032225
2016-03-11    0.032575
2016-03-22    0.032987
2016-03-09    0.033090
2016-03-08    0.033296
2016-04-01    0.033687
2016-03-30    0.033687
2016-03-29    0.034099
2016-03-15    0.034284
2016-03-19    0.034778
2016-03-28    0.034860
2016-04-02    0.035478
2016-03-07    0.036014
2016-04-04    0.036487
2016-03-14    0.036549
2016-03-12    0.036920
2016-03-21    0.037373
2016-03-20    0.037887
2016-04-03    0.038608
Name: date_crawled, dtype: float64

Now moving on to the last seen column of the data

In [23]:
(autos['last_seen'].
                    str[:10].
                    value_counts( normalize = True , dropna = False ).
                    sort_values(ascending=False)
)

2016-04-06    0.221806
2016-04-07    0.131947
2016-04-05    0.124761
2016-03-17    0.028086
2016-04-03    0.025203
2016-04-02    0.024915
2016-03-30    0.024771
2016-04-04    0.024483
2016-03-31    0.023783
2016-03-12    0.023783
2016-04-01    0.022794
2016-03-29    0.022341
2016-03-22    0.021373
2016-03-28    0.020859
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-23    0.018532
2016-03-26    0.016802
2016-03-16    0.016452
2016-03-15    0.015876
2016-03-19    0.015834
2016-03-27    0.015649
2016-03-14    0.012602
2016-03-11    0.012375
2016-03-10    0.010666
2016-03-09    0.009595
2016-03-13    0.008895
2016-03-08    0.007413
2016-03-18    0.007351
2016-03-07    0.005395
2016-03-06    0.004324
2016-03-05    0.001071
Name: last_seen, dtype: float64

As we can see the last three days (april 5, 6, 7) have very high last seen values.
It is not realistic that the sales just jumped and this might have to do with the crawling period ending and might not be actual car sales.


Now getting to the ads created column

In [24]:
(autos['ad_created'].str[:10].
                             value_counts( normalize = True , dropna = False ).
                             sort_index()

)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

We can see that most of the dates on which ads were created are pretty close to the listing dates.
There are also some ads which are pretty old and one dates back to as much as 9 months.

In [25]:
autos['year_of_registration'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: year_of_registration, dtype: float64

The above results show that there is something wrong with the year_of_registration data as the minimum value is 1000 which is not possible and the maximum value is not also possible as the year is 9999 which is far away in the future .


# Dealing with incorrect registration data

The cars were invented in the 1856, but we are going to take the safe approach here and say that the registry may be after the 1900s and unto 2016.
We used unto 2016 as the car cant be registered before the listing was first seen.

Let's see if we can determine the percentage of the values outside the range [1900,2016] and if these end up under 5-7 percent we may just drop it

In [26]:
count= autos['year_of_registration'].between(1900,2016).sum()
print(1-(count/autos.shape[0]))

0.03879336971069702


It is clearly visible that these values are less than 4 percent and we safely just drop these rows

In [27]:
autos = autos[autos['year_of_registration'].between(1900,2016)]

In [28]:
autos['year_of_registration'].value_counts( normalize = True , dropna = False ).head(10)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: year_of_registration, dtype: float64

Now we are going to analyze the brands in the database

In [29]:
autos['brand'].value_counts(dropna= False, normalize= True).sort_values(ascending =False)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

We are now going to aggregate the brands that have at the least 5% contribution to the total

In [40]:
brand_counts= autos['brand'].value_counts(normalize= True)
common_brands = brand_counts[brand_counts>0.05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


Let's make a dictionary to display the average rate of the cars in the common_brands

In [41]:
average_common_brands={}

for brands in common_brands:
    brand= autos[autos['brand']==brands]
    mean_price=brand['price'].mean()
    average_common_brands[brands]=int (mean_price)
print(average_common_brands)

{'volkswagen': 5402, 'bmw': 8332, 'opel': 2975, 'mercedes_benz': 8628, 'audi': 9336, 'ford': 3749}


The Mercedes, BMW and Audi are more expensive amongst the common brands as compared to the Volkswagen, Opel and Ford

# EXPLORING MILEAGE

In [47]:
acb_series = pd.Series(average_common_brands)
brand_count_df=pd.DataFrame(acb_series,columns=['mean_price'])
print(brand_count_df['mean_price'].sort_values(ascending= False))

audi             9336
mercedes_benz    8628
bmw              8332
volkswagen       5402
ford             3749
opel             2975
Name: mean_price, dtype: int64


In [48]:
brand_mean_mileage={}

for brands in common_brands:
    brand=autos[autos['brand']==brands]
    mean_mileage=brand['odometer_km'].mean()
    brand_mean_mileage[brands]=int(mean_mileage)
print(brand_mean_mileage)

{'volkswagen': 128707, 'bmw': 132572, 'opel': 129310, 'mercedes_benz': 130788, 'audi': 129157, 'ford': 124266}


In [49]:
mean_mileage= pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices=pd.Series(average_common_brands).sort_values(ascending=False)


In [50]:
brand_info= pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info

mean_mileage
bmw                  132572
mercedes_benz        130788
opel                 129310
audi                 129157
volkswagen           128707
ford                 124266

In [52]:
brand_info['brand_price']= mean_prices
brand_info

mean_mileage  brand_price
bmw                  132572         8332
mercedes_benz        130788         8628
opel                 129310         2975
audi                 129157         9336
volkswagen           128707         5402
ford                 124266         3749

We can see that the mileage does not vary as much as the prices did by brands. We see only slight variances in the mileage. The brands with higher cost seem to have slightly higher mileage

THIS CONCLUDES OUR PROJECT
# END